File: Longest Valid Parentheses/longest_valid_parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        Stack <Integer> index=new Stack();
        index.push(-1); 
        int max=0;
        for(int i=0; i<s.length();i++){
            if(s.charAt(i)=='('){
                index.push(i);
            }else{ 
                index.pop();
                if(!index.isEmpty()){
                    max=Math.max(max,i-index.peek());
                }else{
                    index.push(i);
                     }
            } 
        }
        return max;
    }
}

File: Longest Valid Parentheses/longest_valid_parentheses.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        stack<char> st;
        stack<int> idx;
        int ans = 0,n = s.length();

        idx.push(-1);

        for(int i=0; i<n; i++){
            if(s[i]=='('){
                st.push('(');
                idx.push(i);
            }
            else{
                if(!st.empty() && st.top()=='('){
                    st.pop();
                    idx.pop();
                    ans = max(ans, i - idx.top());
                }
                else{
                    idx.push(i);
                }
            }
        }

        return ans;
    }
};

int main()
{
    return 0;
}